corehq/warehouse/etl.py

```diff
@@ -0,0 +1,103 @@
+'''
+This files defines ETL objects that all have a common "load" function.
+Each object is meant for transferring one type of data to another type.
+'''
+import os
+
+from django.db import connections
+from django.conf import settings
+from django.template import Context, engines
+
+from corehq.warehouse.utils import django_batch_records
+from corehq.sql_db.routers import db_for_read_write
+
+
+class BaseETLMixin(object):
+
+    @classmethod
+    def load(cls, start_datetime, end_datetime):
+        raise NotImplementedError
+
+
+class CustomSQLETLMixin(BaseETLMixin):
+    '''
+    Mixin for transferring data from a SQL store to another SQL store using
+    a custom SQL script.
+    '''
+
+    @classmethod
+    def load(cls, start_datetime, end_datetime):
+        '''
+        Bulk loads records for a dim or fact table from
+        their corresponding dependencies
+        '''
+        database = db_for_read_write(cls)
+        with connections[database].cursor() as cursor:
+            cursor.execute(cls._sql_query_template(cls.slug, start_datetime, end_datetime))
+
+    @classmethod
+    def _table_context(cls, start_datetime, end_datetime):
+        '''
+        Get a dict of slugs to table name mapping
+        :returns: Dict of slug to table_name
+        {
+            <slug>: <table_name>,
+            ...
+        }
+        '''
+        from corehq.warehouse.models import get_cls_by_slug
+
+        context = {cls.slug: cls._meta.db_table}
+        for dep in cls.dependencies():
```

Comment:
yeah they're not defined on this model, but rather on the warehousemixin. i want all warehouse tables to have dependencies regardless of whether they use custom sql to do their transformation. perhaps the correct thing to do here is to pass in those two variables